File: src/repositories/session-repository.ts
import Database from 'better-sqlite3';
import { Session } from '../domain/models.js';

function mapSession(row: Record<string, unknown>): Session {
  return {
    id: String(row.id),
    userId: String(row.user_id),
    createdAt: String(row.created_at),
  };
}

export interface SessionRepository {
  create(session: Session): void;
  findById(id: string): Session | null;
  deleteById(id: string): void;
}

export function createSessionRepository(db: Database.Database): SessionRepository {
  return {
    create(session) {
      db.prepare(
        `
          INSERT INTO sessions (id, user_id, created_at)
          VALUES (@id, @userId, @createdAt)
        `,
      ).run(session);
    },
    findById(id) {
      const row = db.prepare('SELECT * FROM sessions WHERE id = ?').get(id) as Record<string, unknown> | undefined;
      return row ? mapSession(row) : null;
    },
    deleteById(id) {
      db.prepare('DELETE FROM sessions WHERE id = ?').run(id);
    },
  };
}

File: src/repositories/session_repository.py
import sqlite3

from src.domain.models import Session


def _map_session(row: sqlite3.Row) -> Session:
    return Session(
        id=str(row["id"]),
        user_id=str(row["user_id"]),
        created_at=str(row["created_at"]),
    )


class SessionRepository:
    def __init__(self, connection: sqlite3.Connection) -> None:
        self._connection = connection

    def create(self, session: Session) -> None:
        self._connection.execute(
            "INSERT INTO sessions (id, user_id, created_at) VALUES (?, ?, ?)",
            (session.id, session.user_id, session.created_at),
        )
        self._connection.commit()

    def find_by_id(self, session_id: str) -> Session | None:
        row = self._connection.execute("SELECT * FROM sessions WHERE id = ?", (session_id,)).fetchone()
        return _map_session(row) if row else None

    def delete_by_id(self, session_id: str) -> None:
        self._connection.execute("DELETE FROM sessions WHERE id = ?", (session_id,))
        self._connection.commit()
